README.md
# Rust TCS3400 RGB Color Light to Digital Converter with IR Filter Driver

[![crates.io](https://img.shields.io/crates/v/tcs3400.svg)](https://crates.io/crates/tcs3400)
[![Docs](https://docs.rs/tcs3400/badge.svg)](https://docs.rs/tcs3400)
[![Build Status](https://github.com/andresv/tcs3400-rs/workflows/Build/badge.svg)](https://github.com/andresv/tcs3400-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/andresv/tcs3400-rs/badge.svg?branch=master)](https://coveralls.io/github/andresv/tcs3400-rs?branch=master)

This is a platform agnostic Rust driver for the TCS3400 RGB color light to
digital converter with IR filter, based on the [`embedded-hal`] traits.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal

This driver allows you to:
- Enable/disable the device.
- Enable/disable the RGB converter.
- Set RGB converter gain.
- Enable/disable the RGB converter interrupt generation.
- Set the RGB converter interrupt clear channel low/high thresholds.
- Set the RGB converter interrupt persistence.
- Set the number of integration cycles.
- Enable/disable the wait feature.
- Set the number of wait time cycles.
- Enable/disable the *wait long* setting.
- Read status of RGB converter.
- Read the clear (unfiltered) channel measurement.
- Read the red channel measurement.
- Read the green channel measurement.
- Read the blue channel measurement.
- Read the measurement of all channels at once.
- Read the device ID.

## The device
The TCS3400 device provides a digital return of red, green, blue (RGB), and
clear light sensing values. An IR blocking filter, integrated on-chip and
localized to the color sensing photodiodes, minimizes the IR spectral
component of the incoming light and allows color measurements to be made
accurately. The high sensitivity, wide dynamic range, and IR blocking
filter make the TCS3400 an ideal color sensor solution for use under
varying lighting conditions and through attenuating materials.

The TCS3400 color sensor has a wide range of applications including RGB LED
backlight control, solid-state lighting, health/fitness products,
industrial process controls and medical diagnostic equipment. In addition,
the IR blocking filter enables the TCS3400 to perform ambient light sensing
(ALS). Ambient light sensing is widely used in display-based products such
as cell phones, notebooks, and TVs to sense the lighting environment and
enable automatic display brightness for optimal viewing and power savings.
The TCS3400, itself, can enter a lower-power wait state between light
sensing measurements to further reduce the average power consumption.

Datasheet:
- [TCS3400](https://ams.com/documents/20143/36005/TCS3400_DS000411_5-00.pdf)

This driver is compatible with the devices TCS34005 and TCS34007.

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the device.

```rust
use linux_embedded_hal::I2cdev;
use tcs3400::Tcs3400;

fn main() {
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let mut sensor = Tcs3400::new(dev);
    sensor.enable().unwrap();
    sensor.enable_rgbc().unwrap();
    while !sensor.is_rgbc_status_valid().unwrap() {
        // wait for measurement to be available
    }
    let m = sensor.read_all_channels().unwrap();
    println!(
        "Measurements: clear = {}, red = {}, green = {}, blue = {}",
        m.clear, m.red, m.green, m.blue
    );
}
```

## Minimum Supported Rust Version (MSRV)

This crate is guaranteed to compile on stable Rust 1.31.0 and up. It *might*
compile with older versions but that may change in any new patch release.

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/andresv/tcs3400-rs/issues).

## License

Licensed under either of:

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

* Added `embedded_hal_async` support.

## 0.1.0 - 2022-11-16

This is the initial release to crates.io based on https://github.com/eldruin/tcs3472-rs. All changes will be documented in
this CHANGELOG.

[Unreleased]: https://github.com/andresv/tcs3400-rs/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/andresv/tcs3400-rs/tree/0.1.0


src/types.rs
/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C bus error
    I2C(E),
    /// Invalid input data provided.
    InvalidInputData,
}

/// RGB converter gain
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RgbCGain {
    /// 1x gain
    _1x,
    /// 4x gain
    _4x,
    /// 16x gain
    _16x,
    /// 60x gain
    _60x,
}

/// RGB converter interrupt persistence
///
/// This controls the RGB converter interrupt generation rate.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum RgbCInterruptPersistence {
    /// Every RGBC cycle generates an interrupt.
    Every,
    /// 1 clear channel value out of range.
    Any,
    /// 2 clear channel consecutive values out of range.
    _2,
    /// 3 clear channel consecutive values out of range.
    _3,
    /// 5 clear channel consecutive values out of range.
    _5,
    /// 10 clear channel consecutive values out of range.
    _10,
    /// 15 clear channel consecutive values out of range.
    _15,
    /// 20 clear channel consecutive values out of range.
    _20,
    /// 25 clear channel consecutive values out of range.
    _25,
    /// 30 clear channel consecutive values out of range.
    _30,
    /// 35 clear channel consecutive values out of range.
    _35,
    /// 40 clear channel consecutive values out of range.
    _40,
    /// 45 clear channel consecutive values out of range.
    _45,
    /// 50 clear channel consecutive values out of range.
    _50,
    /// 55 clear channel consecutive values out of range.
    _55,
    /// 60 clear channel consecutive values out of range.
    _60,
}

/// Result of measurement of all channels
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct AllChannelMeasurement {
    /// Red channel measurement.
    pub red: u16,
    /// Green channel measurement.
    pub green: u16,
    /// Blue channel measurement.
    pub blue: u16,
    /// Clear (unfiltered) channel measurement.
    pub clear: u16,
}


src/reading.rs
use crate::{AllChannelMeasurement, BitFlags, Error, Register, Tcs3400, DEVICE_ADDRESS};
use embedded_hal::blocking::i2c;

impl<I2C, E> Tcs3400<I2C>
where
    I2C: i2c::WriteRead<Error = E>,
{
    /// Check whether the RGB converter status is valid.
    ///
    /// Indicates that the RGBC channels have completed an integration cycle.
    #[allow(clippy::wrong_self_convention)]
    pub fn is_rgbc_status_valid(&mut self) -> Result<bool, Error<E>> {
        let status = self.read_register(Register::STATUS)?;
        Ok((status & BitFlags::RGBC_VALID) != 0)
    }

    /// Read the clear (unfiltered) channel measurement data.
    pub fn read_clear_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::CDATAL)
    }

    /// Read the red channel measurement data.
    pub fn read_red_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::RDATAL)
    }

    /// Read the green channel measurement data.
    pub fn read_green_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::GDATAL)
    }

    /// Read the blue channel measurement data.
    pub fn read_blue_channel(&mut self) -> Result<u16, Error<E>> {
        self.read_channel(Register::BDATAL)
    }

    fn read_channel(&mut self, first_register: u8) -> Result<u16, Error<E>> {
        let mut cdata = [0; 2];
        self.read_registers(first_register, &mut cdata)?;
        Ok((u16::from(cdata[1])) << 8 | u16::from(cdata[0]))
    }

    /// Read the measurement data of all channels at once.
    pub fn read_all_channels(&mut self) -> Result<AllChannelMeasurement, Error<E>> {
        let mut data = [0; 8];
        self.read_registers(Register::CDATAL, &mut data)?;
        Ok(AllChannelMeasurement {
            clear: u16::from(data[1]) << 8 | u16::from(data[0]),
            red: u16::from(data[3]) << 8 | u16::from(data[2]),
            green: u16::from(data[5]) << 8 | u16::from(data[4]),
            blue: u16::from(data[7]) << 8 | u16::from(data[6]),
        })
    }

    /// Read the device ID.
    ///
    /// The value returned corresponds to the part number identification:
    /// - `0x90` => `TCS34001 & TCS34005`
    /// - `0x93` => `TCS34003 & TCS34007`
    pub fn read_device_id(&mut self) -> Result<u8, Error<E>> {
        self.read_register(Register::ID)
    }

    fn read_register(&mut self, register: u8) -> Result<u8, Error<E>> {
        let mut data = [0];
        self.i2c
            .write_read(DEVICE_ADDRESS, &[register], &mut data)
            .map_err(Error::I2C)?;
        Ok(data[0])
    }

    fn read_registers(&mut self, first_register: u8, mut data: &mut [u8]) -> Result<(), Error<E>> {
        self.i2c
            .write_read(DEVICE_ADDRESS, &[first_register], &mut data)
            .map_err(Error::I2C)
    }
}


src/interface.rs
pub(crate) const DEVICE_ADDRESS: u8 = 0x39;

pub(crate) struct Register;

impl Register {
    pub(crate) const ENABLE: u8 = 0x80;
    pub(crate) const ATIME: u8 = 0x81;
    pub(crate) const WTIME: u8 = 0x83;
    pub(crate) const AILTL: u8 = 0x84;
    pub(crate) const AILTH: u8 = 0x85;
    pub(crate) const AIHTL: u8 = 0x86;
    pub(crate) const AIHTH: u8 = 0x87;
    pub(crate) const APERS: u8 = 0x8C;
    pub(crate) const CONFIG: u8 = 0x8D;
    pub(crate) const CONTROL: u8 = 0x8F;
    pub(crate) const ID: u8 = 0x92;
    pub(crate) const STATUS: u8 = 0x93;
    pub(crate) const CDATAL: u8 = 0x94;
    pub(crate) const RDATAL: u8 = 0x96;
    pub(crate) const GDATAL: u8 = 0x98;
    pub(crate) const BDATAL: u8 = 0x9A;
}

pub(crate) struct BitFlags;

impl BitFlags {
    pub(crate) const POWER_ON: u8 = 0b0000_0001; // PON
    pub(crate) const RGBC_EN: u8 = 0b0000_0010; // AEN
    pub(crate) const WAIT_EN: u8 = 0b0000_1000; // WEN
    pub(crate) const RGBC_INT_EN: u8 = 0b0001_0000; // AIEN
    pub(crate) const RGBC_VALID: u8 = 0b0000_0001; // AVALID
    pub(crate) const WLONG: u8 = 0b0000_0010;
}


src/configuration.rs
use crate::{
    BitFlags, Error, Register, RgbCGain, RgbCInterruptPersistence, Tcs3400, DEVICE_ADDRESS,
};
use embedded_hal::blocking::i2c;

impl<I2C, E> Tcs3400<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Enable the device (Power ON).
    ///
    /// The device goes to idle state.
    pub fn enable(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable | BitFlags::POWER_ON)
    }

    /// Disable the device (sleep).
    pub fn disable(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable & !BitFlags::POWER_ON)
    }

    /// Enable the RGB converter.
    pub fn enable_rgbc(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable | BitFlags::RGBC_EN)
    }

    /// Disable the RGB converter.
    pub fn disable_rgbc(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable & !BitFlags::RGBC_EN)
    }

    /// Enable the RGB converter interrupt generation.
    pub fn enable_rgbc_interrupts(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable | BitFlags::RGBC_INT_EN)
    }

    /// Disable the RGB converter interrupt generation.
    pub fn disable_rgbc_interrupts(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable & !BitFlags::RGBC_INT_EN)
    }

    /// Enable the wait feature (wait timer).
    pub fn enable_wait(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable | BitFlags::WAIT_EN)
    }

    /// Disable the wait feature (wait timer).
    pub fn disable_wait(&mut self) -> Result<(), Error<E>> {
        let enable = self.enable;
        self.write_enable(enable & !BitFlags::WAIT_EN)
    }

    fn write_enable(&mut self, enable: u8) -> Result<(), Error<E>> {
        self.write_register(Register::ENABLE, enable)?;
        self.enable = enable;
        Ok(())
    }

    /// Set the number of wait time cycles  (1-256).
    ///
    /// The actual wait time depends on the "*wait long*" setting.
    /// - If *wait long* is disabled, then the wait time corresponds to:
    ///   `number_of_cycles * 2.78ms`.
    /// - If *wait long* is enabled, then the wait time is increased by a
    ///   factor of 12 and therefore corresponds to aproximately:
    ///   `number_of_cycles * 0.03s`.
    /// See [`enable_wait_long()`](#method.enable_wait_long) and
    ///  [`disable_wait_long()`](#method.disable_wait_long).
    pub fn set_wait_cycles(&mut self, cycles: u16) -> Result<(), Error<E>> {
        if cycles > 256 || cycles == 0 {
            return Err(Error::InvalidInputData);
        }
        // the value is stored as a two's complement
        self.write_register(Register::WTIME, (256 - cycles as u16) as u8)
    }

    /// Enable the *wait long* setting.
    ///
    /// The wait time configured with `set_wait_cycles()` is increased by a
    /// factor of 12. See [`set_wait_cycles()`](#method.set_wait_cycles).
    pub fn enable_wait_long(&mut self) -> Result<(), Error<E>> {
        self.write_register(Register::CONFIG, BitFlags::WLONG)
    }

    /// Disable the *wait long* setting.
    ///
    /// The wait time configured with `set_wait_cycles()` is used without
    /// multiplication factor. See [`set_wait_cycles()`](#method.set_wait_cycles).
    pub fn disable_wait_long(&mut self) -> Result<(), Error<E>> {
        self.write_register(Register::CONFIG, 0)
    }

    /// Set the RGB converter gain.
    pub fn set_rgbc_gain(&mut self, gain: RgbCGain) -> Result<(), Error<E>> {
        // Register field: AGAIN
        match gain {
            RgbCGain::_1x => self.write_register(Register::CONTROL, 0),
            RgbCGain::_4x => self.write_register(Register::CONTROL, 1),
            RgbCGain::_16x => self.write_register(Register::CONTROL, 2),
            RgbCGain::_60x => self.write_register(Register::CONTROL, 3),
        }
    }

    /// Set the number of integration cycles (1-256).
    ///
    /// The actual integration time corresponds to: `number_of_cycles * 2.78ms`.
    pub fn set_integration_cycles(&mut self, cycles: u16) -> Result<(), Error<E>> {
        if cycles > 256 || cycles == 0 {
            return Err(Error::InvalidInputData);
        }
        // the value is stored as a two's complement
        self.write_register(Register::ATIME, (256 - cycles as u16) as u8)
    }

    /// Set the RGB converter interrupt clear channel low threshold.
    pub fn set_rgbc_interrupt_low_threshold(&mut self, threshold: u16) -> Result<(), Error<E>> {
        self.write_register(Register::AILTL, threshold as u8)?;
        self.write_register(Register::AILTH, (threshold >> 8) as u8)
    }

    /// Set the RGB converter interrupt clear channel high threshold.
    pub fn set_rgbc_interrupt_high_threshold(&mut self, threshold: u16) -> Result<(), Error<E>> {
        self.write_register(Register::AIHTL, threshold as u8)?;
        self.write_register(Register::AIHTH, (threshold >> 8) as u8)
    }

    /// Set the RGB converter interrupt persistence.
    ///
    /// This controls the RGB converter interrupt generation rate.
    pub fn set_rgbc_interrupt_persistence(
        &mut self,
        persistence: RgbCInterruptPersistence,
    ) -> Result<(), Error<E>> {
        use crate::RgbCInterruptPersistence as IP;
        match persistence {
            IP::Every => self.write_register(Register::APERS, 0),
            IP::Any => self.write_register(Register::APERS, 1),
            IP::_2 => self.write_register(Register::APERS, 2),
            IP::_3 => self.write_register(Register::APERS, 3),
            IP::_5 => self.write_register(Register::APERS, 4),
            IP::_10 => self.write_register(Register::APERS, 5),
            IP::_15 => self.write_register(Register::APERS, 6),
            IP::_20 => self.write_register(Register::APERS, 7),
            IP::_25 => self.write_register(Register::APERS, 8),
            IP::_30 => self.write_register(Register::APERS, 9),
            IP::_35 => self.write_register(Register::APERS, 10),
            IP::_40 => self.write_register(Register::APERS, 11),
            IP::_45 => self.write_register(Register::APERS, 12),
            IP::_50 => self.write_register(Register::APERS, 13),
            IP::_55 => self.write_register(Register::APERS, 14),
            IP::_60 => self.write_register(Register::APERS, 15),
        }
    }

    fn write_register(&mut self, register: u8, value: u8) -> Result<(), Error<E>> {
        self.i2c
            .write(DEVICE_ADDRESS, &[register, value])
            .map_err(Error::I2C)
    }
}


src/lib.rs
//! This is a platform agnostic Rust driver for the TCS3400 RGB color light to
//! digital converter with IR filter, based on the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Enable/disable the device.
//! - Enable/disable the RGB converter.
//! - Set RGB converter gain.
//! - Enable/disable the RGB converter interrupt generation.
//! - Set the RGB converter interrupt clear channel low/high thresholds.
//! - Set the RGB converter interrupt persistence.
//! - Set the number of integration cycles.
//! - Enable/disable the wait feature.
//! - Set the number of wait time cycles.
//! - Enable/disable the *wait long* setting.
//! - Read status of RGB converter.
//! - Read the clear (unfiltered) channel measurement.
//! - Read the red channel measurement.
//! - Read the green channel measurement.
//! - Read the blue channel measurement.
//! - Read the measurement of all channels at once.
//! - Read the device ID.
//!
//! ## The device
//!
//! The TCS3400 device provides a digital return of red, green, blue (RGB), and
//! clear light sensing values. An IR blocking filter, integrated on-chip and
//! localized to the color sensing photodiodes, minimizes the IR spectral
//! component of the incoming light and allows color measurements to be made
//! accurately. The high sensitivity, wide dynamic range, and IR blocking
//! filter make the TCS3400 an ideal color sensor solution for use under
//! varying lighting conditions and through attenuating materials.
//!
//! The TCS3400 color sensor has a wide range of applications including RGB LED
//! backlight control, solid-state lighting, health/fitness products,
//! industrial process controls and medical diagnostic equipment. In addition,
//! the IR blocking filter enables the TCS3400 to perform ambient light sensing
//! (ALS). Ambient light sensing is widely used in display-based products such
//! as cell phones, notebooks, and TVs to sense the lighting environment and
//! enable automatic display brightness for optimal viewing and power savings.
//! The TCS3400, itself, can enter a lower-power wait state between light
//! sensing measurements to further reduce the average power consumption.
//!
//! Datasheet:
//! - [TCS3400](https://ams.com/documents/20143/36005/TCS3400_DS000411_5-00.pdf)
//!
//! This driver is compatible with the devices TCS34001 and TCS34003.
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then create an instance of the driver.
//!
//! ### Enable and read the color measurement
//!
//! Import this crate and an `embedded_hal` implementation, then instantiate
//! the device:
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use tcs3400::Tcs3400;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Tcs3400::new(dev);
//! sensor.enable().unwrap();
//! sensor.enable_rgbc().unwrap();
//! while !sensor.is_rgbc_status_valid().unwrap() {
//!     // wait for measurement to be available
//! };
//!
//! let clear = sensor.read_clear_channel().unwrap();
//! let red = sensor.read_red_channel().unwrap();
//! let green = sensor.read_green_channel().unwrap();
//! let blue = sensor.read_blue_channel().unwrap();
//!
//! println!("Measurements: clear = {}, red = {}, green = {}, blue = {}",
//!          clear, red, green, blue);
//! ```
//!
//! ### Read all the channels at once
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use tcs3400::Tcs3400;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Tcs3400::new(dev);
//! sensor.enable().unwrap();
//! sensor.enable_rgbc().unwrap();
//! while !sensor.is_rgbc_status_valid().unwrap() {
//!     // wait for measurement to be available
//! };
//!
//! let measurement = sensor.read_all_channels().unwrap();
//!
//! println!("Measurements: clear = {}, red = {}, green = {}, blue = {}",
//!          measurement.clear, measurement.red, measurement.green,
//!          measurement.blue);
//! ```
//!
//! ### Change the RGB converter gain and integration cycles
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use tcs3400::{RgbCGain, Tcs3400};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Tcs3400::new(dev);
//! sensor.enable().unwrap();
//! sensor.enable_rgbc().unwrap();
//! sensor.set_rgbc_gain(RgbCGain::_16x).unwrap();
//! sensor.set_integration_cycles(32).unwrap();
//! ```
//!
//! ### Enable wait function and set wait time to 1.008s
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use tcs3400::Tcs3400;
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Tcs3400::new(dev);
//! sensor.enable().unwrap();
//! sensor.enable_rgbc().unwrap();
//! // This results in 35 * 2.4ms * 12 = 1.008s
//! sensor.set_wait_cycles(35).unwrap();
//! sensor.enable_wait_long().unwrap(); // 12x mutiplicator
//! sensor.enable_wait().unwrap(); // actually enable wait timer
//! ```
//!
//! ### Enable and configure RGB converter interrupt generation
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use tcs3400::{RgbCInterruptPersistence, Tcs3400};
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Tcs3400::new(dev);
//! sensor.enable().unwrap();
//! sensor.enable_rgbc().unwrap();
//! sensor.set_rgbc_interrupt_low_threshold(1024).unwrap();
//! sensor.set_rgbc_interrupt_high_threshold(61440).unwrap();
//! sensor.set_rgbc_interrupt_persistence(RgbCInterruptPersistence::_5).unwrap();
//! sensor.enable_rgbc_interrupts().unwrap();
//! ```

#![deny(unsafe_code, missing_docs)]
#![no_std]

use embedded_hal::blocking::i2c;

mod configuration;
mod interface;
use crate::interface::{BitFlags, Register, DEVICE_ADDRESS};
mod reading;
mod types;
pub use crate::types::{AllChannelMeasurement, Error, RgbCGain, RgbCInterruptPersistence};

/// TCS3400 device driver.
#[derive(Debug)]
pub struct Tcs3400<I2C> {
    /// The concrete I²C device implementation.
    i2c: I2C,
    /// Enable register status
    enable: u8,
}

impl<I2C, E> Tcs3400<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Create new instance of the TCS3400 device.
    pub fn new(i2c: I2C) -> Self {
        Tcs3400 { i2c, enable: 0 }
    }

    /// Destroy driver instance, return I²C bus instance.
    pub fn destroy(self) -> I2C {
        self.i2c
    }
}
